libbpf-tools/runqlen.c

```diff
@@ -0,0 +1,283 @@
+"\n"
+"EXAMPLES:\n"
+"    runqlen         # summarize run queue length as a histogram\n"
+"    runqlen 1 10    # print 1 second summaries, 10 times\n"
+"    runqlen -T 1    # 1s summaries and timestamps\n"
+"    runqlen -O      # report run queue occupancy\n"
+"    runqlen -C      # show each CPU separately\n";
+
+static const struct argp_option opts[] = {
+	{ "cpus", 'C', NULL, 0, "Print output for each CPU separately" },
+	{ "runqocc", 'O', NULL, 0, "Report run queue occupancy" },
+	{ "timestamp", 'T', NULL, 0, "Include timestamp on output" },
+	{ "verbose", 'v', NULL, 0, "Verbose debug output" },
+	{},
+};
+
+static error_t parse_arg(int key, char *arg, struct argp_state *state)
+{
+	static int pos_args;
+
+	switch (key) {
+	case 'v':
+		env.verbose = true;
+		break;
+	case 'C':
+		env.per_cpu = true;
+		break;
+	case 'O':
+		env.runqocc = true;
+		break;
+	case 'T':
+		env.timestamp = true;
+		break;
+	case ARGP_KEY_ARG:
+		errno = 0;
+		if (pos_args == 0) {
+			env.interval = strtol(arg, NULL, 10);
+			if (errno) {
+				fprintf(stderr, "invalid internal\n");
+				argp_usage(state);
+			}
+		} else if (pos_args == 1) {
+			env.times = strtol(arg, NULL, 10);
+			if (errno) {
+				fprintf(stderr, "invalid times\n");
+				argp_usage(state);
+			}
+		} else {
+			fprintf(stderr,
+				"unrecognized positional argument: %s\n", arg);
+			argp_usage(state);
+		}
+		pos_args++;
+		break;
+	default:
+		return ARGP_ERR_UNKNOWN;
+	}
+	return 0;
+}
+
+static int nr_cpus;
+
+static int open_and_attach_perf_event(int freq, struct bpf_program *prog,
+				struct bpf_link *links[])
+{
+	struct perf_event_attr attr = {
+		.type = PERF_TYPE_SOFTWARE,
+		.freq = 1,
+		.sample_period = freq,
+		.config = PERF_COUNT_SW_CPU_CLOCK,
+	};
+	int i, fd;
+
+	for (i = 0; i < nr_cpus; i++) {
+		fd = syscall(__NR_perf_event_open, &attr, -1, i, -1, 0);
+		if (fd < 0) {
+			fprintf(stderr, "failed to init perf sampling: %s\n",
+				strerror(errno));
+			return -1;
+		}
+		links[i] = bpf_program__attach_perf_event(prog, fd);
+		if (libbpf_get_error(links[i])) {
+			fprintf(stderr, "failed to attach perf event on cpu: "
+				"%d\n", i);
+			links[i] = NULL;
+			close(fd);
+			return -1;
+		}
+	}
+
+	return 0;
+}
+
+int libbpf_print_fn(enum libbpf_print_level level,
+		    const char *format, va_list args)
+{
+	if (level == LIBBPF_DEBUG && !env.verbose)
+		return 0;
+	return vfprintf(stderr, format, args);
+}
+
+static void sig_handler(int sig)
+{
+	exiting = true;
+}
+
+static struct hist zero;
+
+static void print_runq_occupancy(struct runqlen_bpf__bss *bss)
+{
+	__u64 samples, idle = 0, queued = 0;
+	struct hist hist;
+	int slot, i = 0;
+	float runqocc;
+
+	do {
+		hist = bss->hists[i];
+		bss->hists[i] = zero;
+		for (slot = 0; slot < MAX_SLOTS; slot++) {
+			__u64 val = hist.slots[slot];
+
+			if (slot == 0)
+				idle += val;
+			else
+				queued += val;
+		}
+		samples = idle + queued;
+		if (samples)
+			runqocc = queued * 1.0 / samples;
```

Comment:
what if `samples` is zero?

Reply:
I'm not sure whether perf event may fail to sample under a few special circumstances or not?
Is it better to print a waring info or error info then exit instead of setting `runqocc` to zero in else branch?

Reply:
better safe than sorry, seems like it's theoretically possible to have some CPU histograms as all zeros. I'd just do something like `runqocc = queued * 1.0 / max(1, samples);` to be on the safer side. No need for warning message.

Reply:
> better safe than sorry

I love this!

> seems like it's theoretically possible to have some CPU histograms as all zeros.

I am curious about the theoretically,  would you please share me some more details? Thanks :)

> I'd just do something like runqocc = queued * 1.0 / max(1, samples); to be on the safer side. No need for warning message.

Changed to this, thank you！